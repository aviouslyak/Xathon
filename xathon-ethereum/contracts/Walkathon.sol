// SPDX-License-Identifier: MIT
pragma solidity ^0.8.10;

contract XathonFactory {
    struct KeyMapping {
      mapping(string => Xathon) map; 
      string[] keys; 
    }

    KeyMapping private deployedXathons; 
    mapping(string => bool) public takenNames; 
    
    function deployXathon(address to, uint minimum, string memory unit, string memory name, string memory description) public {
        require(!takenNames[name], "Name cannot be taken");
        Xathon newXathon = new Xathon(to, minimum, unit, name, description); 
        takenNames[name] = true; 
        
        deployedXathons.map[name] = newXathon; 
        deployedXathons.keys.push(name); 
    }
    
    function getDeployedXathons() public view returns (string[] memory) {
        return deployedXathons.keys; 
    }

    function getAddress(string memory name) public view returns (Xathon) {
      require(takenNames[name], "Contract with name must exist"); 
      return deployedXathons.map[name];
    }
}

contract Xathon {

  struct Donation {
    uint maxValue;
    uint valuePerUnit;
    uint paid;
  }

  struct KeyMapping {
    mapping(address => Donation) map;
    address[] keys;
  }

  address public recipient;
  string public xathonName;
  string public xathonDescription; 

  // the minimum amount of units a contributor must pay for
  uint public minimumUnitDeposit;
  uint public donationPerUnit; 
  string public xathonUnit; 
  KeyMapping private donations;
                 

  event MoneyToClaim(address claimer, uint remaining);
  
  constructor(address to, uint minimum, string memory unit, string memory name, string memory description){
    minimumUnitDeposit = minimum;
    recipient = to;
    xathonName = name; 
    xathonDescription = description;
    xathonUnit = unit;
    donationPerUnit = 0; 
  }

  function contribute(uint valuePerUnit) public payable {
    // require to pay for at least minimumUnitDeposit amount of units
    require(valuePerUnit * minimumUnitDeposit <= msg.value, "Must pay for a minimum amount of units");
    donations.map[msg.sender] = Donation({
      maxValue: msg.value,
      valuePerUnit: valuePerUnit,
      paid: 0
    });
    donations.keys.push(msg.sender);
    donationPerUnit += valuePerUnit; 
  }

  function pay(Donation storage d, uint walked) private {
    require(msg.sender == recipient);
    uint toPay = d.valuePerUnit * walked;
    if (toPay > d.maxValue) {
      toPay = d.maxValue;
 
    }
    (bool sent, ) = recipient.call{value: toPay}("");
    d.paid = toPay;
    donationPerUnit -= d.valuePerUnit; 
    require(sent, "Ether failed to send");
    if (toPay < d.maxValue) {
      emit MoneyToClaim(msg.sender, d.maxValue - toPay);
    }
  }

  function collectFunds(uint walked) public {
    require(msg.sender == recipient, "Only the recipient can collect the funds");
    for (uint i = 0; i < donations.keys.length; i++) {
      pay(donations.map[donations.keys[i]], walked);
    }
  }

  function claimUnspentFunds() public {
    Donation storage donation = donations.map[msg.sender];
    uint remainingToPay = donation.maxValue - donation.paid;
    require(remainingToPay > 0, "Must have remaining money to collect");
    (bool sent, ) = msg.sender.call{value: remainingToPay}("");
    require(sent, "Ether failed to send");
  }
}
