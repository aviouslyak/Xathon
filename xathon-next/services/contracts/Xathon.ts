import { Xathon as XathonType } from "../../types/web3-v1-contracts/Xathon";
import { TransactionReceipt } from "web3-core";
import { getWeb3Writer, web3ReadOnly } from "../web3";

const contractJson = require("../../build/contracts/Xathon.json");
const abi = contractJson.abi;

class Xathon {
  private xathonContract: XathonType;
  private addresses: string[];

  constructor(address: string, readOnly = true) {
    this.web3 = getWeb3Writer();
    this.ReadOnlyX = web3ReadOnly.eth.Contract;
    this.xathonContract = new web3.eth.Contract(
      abi,
      address
    ) as any as XathonType;
    this.addresses = [];
  }

  async getAddresses(): Promise<void> {
    this.addresses = await web3.eth.requestAccounts();
  }
  async getName(): Promise<string> {
    return await this.xathonContract.methods.xathonName().call();
  }

  async getUnit(): Promise<string> {
    return await this.xathonContract.methods.xathonUnit().call();
  }

  async getDonationPerUnit(): Promise<string> {
    return await this.xathonContract.methods.donationPerUnit().call();
  }

  async getMinimumDeposit(): Promise<string> {
    return await this.xathonContract.methods.minimumUnitDeposit().call();
  }

  async getDescription(): Promise<string> {
    return await this.xathonContract.methods.xathonDescription().call();
  }

  async getRecipient(): Promise<string> {
    return await this.xathonContract.methods.recipient().call();
  }

  async contribute(
    valuePerUnit: number,
    value: number
  ): Promise<TransactionReceipt> {
    return await this.xathonContract.methods.contribute(valuePerUnit).send({
      from: this.addresses[0],
      value: web3.utils.toWei(value.toString(), "ether"),
    });
  }

  async collectFunds(walked: number): Promise<TransactionReceipt> {
    return await this.xathonContract.methods.collectFunds(walked).send({
      from: this.addresses[0],
    });
  }

  async claimUnspentFunds(walked: number): Promise<TransactionReceipt> {
    return await this.xathonContract.methods.claimUnspentFunds().send({
      from: this.addresses[0],
    });
  }
}

export default Xathon;
