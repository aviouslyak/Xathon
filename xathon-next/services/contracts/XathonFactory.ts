import { web3ReadOnly } from "../web3";
import { XathonFactory as XathonFactoryType } from "../../types/web3-v1-contracts/XathonFactory";
const contractJSON = require("../../ethereum/contracts/XathonFactory.json");

const ADDRESSES = {
  rinkeby: "0x7d6a89f57cF666e613baA13a1cce354237b070b0",
};

export default class XathonFactory {
  static ReadOnlyXF: XathonFactoryType = new web3ReadOnly.eth.Contract(
    contractJSON.abi,
    ADDRESSES.rinkeby
  ) as any as XathonFactoryType;

  private WriteOnlyXF: XathonFactoryType;
  constructor() {
    this.WriteOnlyXF = null as any as XathonFactoryType;
  }
  static async getContractNames() {
    return await this.ReadOnlyXF.methods.getDeployedXathons().call();
  }

  static async getContractAddress(name: string) {
    return await this.ReadOnlyXF.methods.getAddress(name).call();
  }
}
