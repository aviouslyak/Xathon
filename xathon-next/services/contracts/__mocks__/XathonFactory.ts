const ADDRESS = "0x0000000000000000000000000000000000000000";

export default class XathonFactory {
  static async getContractNames() {
    return await new Promise((resolve) => resolve(["foo", "boo", "zoo"]));
  }

  static async getContractAddress(name: string) {
    return await new Promise((resolve) => resolve(ADDRESS));
  }
}
