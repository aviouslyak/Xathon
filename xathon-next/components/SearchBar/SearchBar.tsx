import React, { useEffect } from "react";
import { AiOutlineSearch } from "react-icons/ai";
import Input from "../Input/Input";
import Dropdown from "../Dropdown/Dropdown";
import Router from "next/router";
import { XathonFactoryReader } from "../../services/contracts/XathonFactory";

interface Props {
  queryItems: string[];
}

const SearchBar: React.FC<Props> = ({ queryItems }) => {
  const [dropdownActive, setDropdownActive] = React.useState(false);
  const [query, setQuery] = React.useState("");
  const [returnedItems, setReturnedItems] = React.useState<string[]>([]);
  const [error, setError] = React.useState("");

  useEffect(() => {
    // if address, set as address
    if (query.startsWith("0x")) {
      setReturnedItems([query]);
      // filter items to items that start with query
    } else {
      setReturnedItems(
        queryItems.filter((item) => {
          return item.startsWith(query) && query !== "";
        })
      );
    }
  }, [query, queryItems]);

  useEffect(() => {
    // if only one item in dropdown && that item is the query, don't display dropdown
    if (
      returnedItems.length === 1 &&
      !query.startsWith("0x") &&
      query === returnedItems[0]
    ) {
      setDropdownActive(false);
    } else if (returnedItems.length > 0 || query.startsWith("0x")) {
      setDropdownActive(true);
    } else {
      setDropdownActive(false);
    }

    setError("");
  }, [query, returnedItems]);

  const handleOnBlur = () => {
    setTimeout(() => {
      setDropdownActive(false);
    }, 200);
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const xathonFactoryReader = new XathonFactoryReader();
    const match = /^0x[a-fA-F0-9]{40}$/;
    if (match.test(query)) {
      Router.push(`/${query}`);
      return;
    }

    try {
      const address = await xathonFactoryReader.getContractAddress(query);
      address && Router.push(`/${address}`);
    } catch (err: any) {
      setError("Contract not found");
      return;
    }
  };

  return (
    <form onSubmit={handleSubmit} className="w-1/2 text-center">
      <label htmlFor="search" className="font-bold text-2xl">
        Search
      </label>
      <Input
        name="search"
        type="search"
        showLabel={false}
        onChange={(e) => setQuery(e.target.value)}
        value={query}
        dropdownActive={dropdownActive}
        onBlur={handleOnBlur}
        placeholder="Search for a xathon"
        errorText={error}
        id="search"
        icon={
          <button
            data-testid="submit-button"
            className="block m-auto"
            type="submit"
          >
            <AiOutlineSearch className="transition-transform transform hover:scale-125" />
          </button>
        }
        width=""
      />
      {dropdownActive && (
        <Dropdown returnedItems={returnedItems} setQuery={setQuery} />
      )}
    </form>
  );
};

export default SearchBar;
