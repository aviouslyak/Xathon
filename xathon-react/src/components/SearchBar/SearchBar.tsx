import React, { useEffect } from "react";
import { AiOutlineSearch } from "react-icons/ai";
import Input from "../Input/Input";
import Dropdown from "../Dropdown/Dropdown";
import XathonFactory from "../../services/contracts/xathonFactory";

interface Props {
  queryItems: string[];
  setQueryItems: React.Dispatch<React.SetStateAction<string[]>>;
}
const SearchBar: React.FC<Props> = ({ queryItems, setQueryItems }) => {
  const [dropdownActive, setDropdownActive] = React.useState(false);
  const [query, setQuery] = React.useState("");
  const [returnedItems, setReturnedItems] = React.useState<string[]>([]);

  useEffect(() => {
    const getXathons = async () => {
      const xathons = await XathonFactory.getDeployedXathons();
      xathons && setQueryItems(xathons);
    };
    getXathons();
  }, [setQueryItems]);

  useEffect(() => {
    if (query.startsWith("0x")) {
      setReturnedItems([query]);
    } else {
      setReturnedItems(
        queryItems.filter((item) => {
          return item.startsWith(query) && query !== "";
        })
      );
    }
  }, [query, queryItems]);

  useEffect(() => {
    if (returnedItems.includes(query) && !query.startsWith("0x")) {
      setDropdownActive(false);
    } else if (returnedItems.length > 0) {
      setDropdownActive(true);
    } else {
      setDropdownActive(false);
    }
  }, [query, returnedItems]);

  const handleOnBlur = () => {
    setTimeout(() => {
      setDropdownActive(false);
    }, 200);
  };
  return (
    <>
      <label htmlFor="search" className="text-lg font-bold">
        Search for a <span className="accent-text">Xathon</span>
      </label>
      <Input
        name="search"
        type="search"
        showLabel={false}
        onChange={(e) => setQuery(e.target.value)}
        value={query}
        dropdownActive={dropdownActive}
        onBlur={handleOnBlur}
        icon={
          <AiOutlineSearch className="block m-auto transition-transform hover:text-gray-500 transform hover:scale-110" />
        }
        width="responsive-component-width"
      />
      {dropdownActive && (
        <Dropdown returnedItems={returnedItems} setQuery={setQuery} />
      )}
    </>
  );
};

export default SearchBar;
