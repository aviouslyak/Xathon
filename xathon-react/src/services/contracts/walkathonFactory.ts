import web3 from "../web3";
import { AbiItem } from "web3-utils";
import { XathonFactory } from "../../types/web3-v1-contracts/XathonFactory";

const addresses = {
  localhost: "0x80dbA3D7D54ea6dA24737eCa742d7a2F29Df8AE8",
  rinkeby: "0xE64DC9d21522f54783Ede9cA54D2Ab311b37F3De",
};

const abi: AbiItem[] = [
  {
    inputs: [
      {
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "minimum",
        type: "uint256",
      },
      {
        internalType: "string",
        name: "name",
        type: "string",
      },
      {
        internalType: "string",
        name: "description",
        type: "string",
      },
    ],
    name: "deployWalkathon",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "string",
        name: "name",
        type: "string",
      },
    ],
    name: "getAddress",
    outputs: [
      {
        internalType: "contract Walkathon",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "getDeployedWalkathons",
    outputs: [
      {
        internalType: "string[]",
        name: "",
        type: "string[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "string",
        name: "",
        type: "string",
      },
    ],
    name: "takenNames",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

interface WalkathonValues {
  address: string;
  name: string;
  description: string;
  minimumDeposit: number;
}

class WalkathonFactory {
  private static wFactory = new web3.eth.Contract(
    abi,
    addresses["rinkeby"]
  ) as any as XathonFactory;

  static async getDeployedXathons() {
    return await this.wFactory.methods.getDeployedXathons().call();
  }

  static async deployXathon(values: WalkathonValues) {
    const addresses = await web3.eth.getAccounts();
    return await this.wFactory.methods
      .deployXathon(
        values.address,
        values.minimumDeposit,
        "miles",
        values.name,
        values.description
      )
      .send({ from: addresses[0] });
  }
}
export default WalkathonFactory;
