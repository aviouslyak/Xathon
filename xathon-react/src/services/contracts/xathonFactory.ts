import web3 from "../web3";
import { AbiItem } from "web3-utils";
import { XathonFactory } from "../../types/web3-v1-contracts/XathonFactory";

const addresses = {
  localhost: "0x80dbA3D7D54ea6dA24737eCa742d7a2F29Df8AE8",
  rinkeby: "0x7d6a89f57cF666e613baA13a1cce354237b070b0",
};

const abi: AbiItem[] = [
  {
    inputs: [
      {
        internalType: "string",
        name: "",
        type: "string",
      },
    ],
    name: "takenNames",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "minimum",
        type: "uint256",
      },
      {
        internalType: "string",
        name: "unit",
        type: "string",
      },
      {
        internalType: "string",
        name: "name",
        type: "string",
      },
      {
        internalType: "string",
        name: "description",
        type: "string",
      },
    ],
    name: "deployXathon",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "getDeployedXathons",
    outputs: [
      {
        internalType: "string[]",
        name: "",
        type: "string[]",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "string",
        name: "name",
        type: "string",
      },
    ],
    name: "getAddress",
    outputs: [
      {
        internalType: "contract Xathon",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

interface XathonValues {
  address: string;
  name: string;
  description: string;
  unit: string;
  minimumDeposit: number;
}

class XathonFactory {
  private static wFactory = new web3.eth.Contract(
    abi,
    addresses["rinkeby"]
  ) as any as XathonFactory;

  static async getDeployedXathons() {
    return await this.wFactory.methods.getDeployedXathons().call();
  }

  static async deployXathon(values: XathonValues) {
    const addresses = await web3.eth.requestAccounts();
    return await this.wFactory.methods
      .deployXathon(
        values.address,
        values.minimumDeposit,
        values.unit,
        values.name,
        values.description
      )
      .send({ from: addresses[0] });
  }
}
export default XathonFactory;
