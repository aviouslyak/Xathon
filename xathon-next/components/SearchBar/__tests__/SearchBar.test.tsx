import {
  fireEvent,
  render,
  screen,
  waitFor,
  act,
} from "@testing-library/react";
import React from "react";
import { XathonFactoryReader } from "../../../services/contracts/XathonFactory";
import SearchBar from "../SearchBar";
import Router from "next/router";

const MockSearchBar: React.FC = () => {
  const queryItems = ["foo", "boo", "zoo"];
  return <SearchBar queryItems={queryItems} />;
};

const ADDRESS = "0x0000000000000000000000000000000000000000";
jest.mock("../../../services/contracts/XathonFactory");

describe("<SearchBar />", () => {
  let component: HTMLElement;
  let mockPush: jest.SpyInstance;
  let mockGetContractAddress: jest.SpyInstance;

  beforeEach(async () => {
    render(<MockSearchBar />);
    component = screen.getByRole("searchbox");
    mockGetContractAddress = jest
      .spyOn(XathonFactoryReader.prototype, "getContractAddress")
      .mockImplementation(() => new Promise((resolve) => resolve(ADDRESS)));
    mockPush = jest.spyOn(Router, "push");
  });

  test("matches snapshot", () => {
    expect(component).toMatchSnapshot();
  });

  test("dropdown does not display on default", () => {
    const dropdownContainer = screen.queryByTestId("dropdown-container");
    expect(dropdownContainer).not.toBeInTheDocument();
  });

  test("displays dropdown when searching", () => {
    fireEvent.change(component, { target: { value: "f" } });
    const dropdownContainer = screen.getByTestId("dropdown-container");

    expect(dropdownContainer).toBeInTheDocument();
  });

  test("updates search query when clicking dropdown", () => {
    fireEvent.change(component, { target: { value: "f" } });
    const button = screen.getByText("foo");
    fireEvent.click(button);
    expect(component).toHaveDisplayValue("foo");
  });

  test("blurring input closes dropdown", async () => {
    fireEvent.change(component, { target: { value: "f" } });
    await act(async () => {
      fireEvent.blur(component);
      await new Promise((resolve) => setTimeout(resolve, 200));
    });
  });

  test("dropdown displays one item when entering contract", () => {
    fireEvent.change(component, { target: { value: "f" } });
    const dropdownContainer = screen.getByTestId("dropdown-container");

    expect(dropdownContainer.children.length).toBe(1);
  });

  test("calls Router.push on submit", async () => {
    fireEvent.change(component, { target: { value: "foo" } });
    const submitButton = screen.getByRole("button");
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(mockPush).toHaveBeenCalled();
      expect(mockPush).toHaveBeenCalledWith(`/${ADDRESS}`);
    });
  });

  test("doesn't call Router.push on submit failure", async () => {
    fireEvent.change(component, { target: { value: "f" } });
    const submitButton = screen.getByTestId("submit-button");
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(mockPush).not.toHaveBeenCalled();
    });
  });

  test("sets errorText on failure", async () => {
    mockGetContractAddress.mockRejectedValueOnce("");
    fireEvent.change(component, { target: { value: "f" } });
    const submitButton = screen.getByTestId("submit-button");
    fireEvent.click(submitButton);

    await waitFor(() => {
      const errorElement = screen.getByText("Contract not found");
      expect(errorElement).toBeInTheDocument();
    });
  });
});
