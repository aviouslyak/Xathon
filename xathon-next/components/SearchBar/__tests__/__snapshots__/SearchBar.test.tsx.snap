// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> matches snapshot 1`] = `
<input
  class="rounded-lg bg-pink-200 focus:outline-none focus:border-2 border-pink-500 shadow-md shadow-pink-800  h-full w-full pl-2 pr-10  "
  id="search"
  name="search"
  placeholder="Search for a xathon"
  type="search"
  value=""
/>
`;
